### PR TITLE
Error handling with alerts

### PR DESCRIPTION
**Pull request by crustypluto19**
- Added error handling for exercise view
- Refactored failed authentication (the previous implementation of the invalid credentials error wasn't working for me consistently)

https://user-images.githubusercontent.com/24537483/215289335-16d7286a-3b05-4e8f-bce3-baa36676f53c.mp4

It is now possible to extend any view to show an error alert, for this however an error property needs to be kept by the corresponding view models.

### CODE REVIEW

**Comment by botrufus** [comment]
Generally a good addition especially for all views except the authentication view. I think it would be nicer to have a red error text below the textfields, as its also common practice amongst other applications,

**Comment by crustypluto19** [comment]
> Generally a good addition especially for all views except the authentication view. I think it would be nicer to have a red error text below the textfields, as its also common practice amongst other applications,

As requested, I added a red border and some animation to indicate that the input might have been false.


https://user-images.githubusercontent.com/24537483/215339084-fb3d3ce9-a921-42b4-971c-8f2f32d546bc.mp4

File: Themis/Views/Authentication/AuthenticationView.swift [hunk 1]
```diff
@@ -57,15 +59,23 @@ struct AuthenticationView: View {
...
     }
 
     var authenticateButton: some View {
         Button {
             Task {
                 await authenticationVM.authenticate()
+                if authenticationVM.error != nil {
+                    withAnimation(.linear) {
+                        self.invalidAttempts += 1
+                        print(invalidAttempts)
```
Comment (pal03377) [inline Themis/Views/Authentication/AuthenticationView.swift]: This print could be removed

File: Themis/Views/Authentication/AuthenticationView.swift [hunk 2]
```diff
@@ -136,3 +164,19 @@ struct LoginButtonStyle: ButtonStyle {
...
+    }
+    
+    init(animatableData: CGFloat) {
+       position = animatableData
+   }
+    
+    func effectValue(size: CGSize) -> ProjectionTransform {
+        ProjectionTransform(CGAffineTransform(translationX: -5 * sin(position * 2 * .pi), y: 0))
+    }
+}
```
Comment (pal03377) [inline Themis/Views/Authentication/AuthenticationView.swift]: Can we put this into another file?

File: Themis/Views/Authentication/AuthenticationView.swift [hunk 3]
```diff
@@ -121,9 +139,19 @@ struct LoginTextFieldStyle: TextFieldStyle {
...
+                    RoundedRectangle(cornerRadius: 10)
+                        .stroke(.red.opacity(0.4), lineWidth: 2)
+                }
+            }
     }
 }
 
+enum TextFieldType {
+    case serverURL, username
+}
```
Comment (pal03377) [inline Themis/Views/Authentication/AuthenticationView.swift]: Considering that in Swift, the names are not scoped to the file by default and just globally accessible: Maybe we should call this `AuthTextFieldType` to not potentially interfere with later additions to the app?

File: Themis/Views/Exercises/ExerciseView.swift [hunk 4]
```diff
@@ -90,7 +91,6 @@ struct ExerciseView: View {
         Button {
             Task {
                 await assessmentVM.initRandomSubmission(exerciseId: exercise.id)
-                UndoManagerSingleton.shared.undoManager.removeAllActions()
```
Comment (pal03377) [inline Themis/Views/Exercises/ExerciseView.swift:90]: Why did you remove this? It seems unrelated to errors.

**Review by pal03377 — APPROVED** [review]
Nice! I'm happy that you did this kind of a chore to do some proper error handling. 🎉 

That said, there are still 13 results when searching for `print(error` in the project. Maybe we could extend this PR to other places as well?

File: Themis/Views/Exercises/ExerciseView.swift (hunk 4 above)
Comment (crustypluto19) [reply Themis/Views/Exercises/ExerciseView.swift:90]: Yes I decided to remove this here as it somehow got left there, it should've been removed before when refactoring the undo/redo. Creating a new PR for this just seems a bit overkill 😅

File: Themis/Views/Authentication/AuthenticationView.swift (hunk 3 above)
Comment (crustypluto19) [reply Themis/Views/Authentication/AuthenticationView.swift]: Good point!

File: Themis/Views/Authentication/AuthenticationView.swift (hunk 1 above)
Comment (crustypluto19) [reply Themis/Views/Authentication/AuthenticationView.swift]: Old habits die hard I guess 😅

**Comment by crustypluto19** [comment]
Latest commit includes:
- Dismissing the view on error (for now only in Exercise View, as shown in the video)
- Removed `print(error`s (only 5 left now, not sure how to best handle the remaining ones)

https://user-images.githubusercontent.com/24537483/215627259-1433159b-1ea3-4f65-b8ba-b62f0a8a049e.mp4


I commented out the error handling in the CourseView for now as I always get "forbidden" for some reason, as shown below:
Simulator Screen Shot - iPad Pro (11-inch) (4th generation) - 2023-01-31 at 01 14 00
Apparently fetchAllCourses in the CourseViewModel always returns a forbidden for some reason.
image

**Comment by pal03377** [comment]
Idea: For the "empty response" errors when starting a new assessment, we could instead explain that there are probably no submissions left to assess.

**Comment by crustypluto19** [comment]
> Idea: For the "empty response" errors when starting a new assessment, we could instead explain that there are probably no submissions left to assess.

Like so?
simulator_screenshot_15971727-8014-4D06-9118-09461F1D024A

**Review by katjanakosic — APPROVED** [review]
Looks good to me 👍🏻
